Bootstrap command refactoring: make command output API more bulletproof (step 7)

Kobzol opened the pull request:
Continuation of https://github.com/rust-lang/rust/pull/127680.

This PR modifies the API of running commands to make it more explicit when a command is expected to produce programmatically handled output. Now if you call just `run`, you cannot access the stdout/stderr by accident, because it will not be returned to the caller.

This API change might be seen as overkill, let me know what do you think. In any case, I'd like to land the second commit, to make it harder to accidentally read stdout/stderr of commands that did not capture output (now you'd get an empty string as a result, but you should probably get a panic instead, if you try to read uncaptured stdout/stderr). 

Tracking issue: https://github.com/rust-lang/rust/issues/126819

r? @onur-ozkan

try-job: x86_64-msvc

Kobzol committed:
Make command output capturing more explicit

Now there are separate functions for running a command without capturing, running while capturing stdout
and running while capturing everything. This should help avoid situations where stdout/stderr is accessed
when it was not captured.

Kobzol committed:
Make it easier to detect when bootstrap tries to read uncaptured stdout/stderr

If e.g. only stdout is captured, but the caller tries to read stderr, previously
they would get back an empty string. Now the code will explicitly panic when
accessing an uncaptured output stream.

bors commented:
:umbrella: The latest upstream changes (presumably #127924) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

onur-ozkan commented:
Didn't see anything wrong.

@bors r+

bors commented:
:pushpin: Commit 1984a463df694e40b3415d7beb50f414a3a78f49 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 1984a463df694e40b3415d7beb50f414a3a78f49 -->
<!-- homu: {"type":"Approved","sha":"1984a463df694e40b3415d7beb50f414a3a78f49","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors rollup=never

bors commented:
:hourglass: Testing commit 1984a463df694e40b3415d7beb50f414a3a78f49 with merge 9e82096798a8089c8dd503fadc57459e3d313799...
<!-- homu: {"type":"BuildStarted","head_sha":"1984a463df694e40b3415d7beb50f414a3a78f49","merge_sha":"9e82096798a8089c8dd503fadc57459e3d313799"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10052589072/job/27784107376)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10052589072/job/27784107376","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10052589072/job/27784035148) [(plain)](https://github.com/rust-lang-ci/rust/commit/9e82096798a8089c8dd503fadc57459e3d313799/checks/27784035148/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling home v0.5.9
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
---
   Compiling bootstrap v0.0.0 (/Users/runner/work/rust/rust/src/bootstrap)
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
---
   Compiling bootstrap v0.0.0 (/Users/runner/work/rust/rust/src/bootstrap)
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
---
   Compiling bootstrap v0.0.0 (/Users/runner/work/rust/rust/src/bootstrap)
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
---
   Compiling bootstrap v0.0.0 (/Users/runner/work/rust/rust/src/bootstrap)
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
```

</details>

Kobzol committed:
Fix usages of old command API

Kobzol commented:
Rebased.

@bors r=onur-ozkan

bors commented:
:pushpin: Commit 037b6261f897a516412f527e980d67eb9acf1413 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 037b6261f897a516412f527e980d67eb9acf1413 -->
<!-- homu: {"type":"Approved","sha":"037b6261f897a516412f527e980d67eb9acf1413","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 037b6261f897a516412f527e980d67eb9acf1413 with merge 18bebddbbdce6814465c533c7f555330b7580293...
<!-- homu: {"type":"BuildStarted","head_sha":"037b6261f897a516412f527e980d67eb9acf1413","merge_sha":"18bebddbbdce6814465c533c7f555330b7580293"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10078545117/job/27863702005) [(plain)](https://github.com/rust-lang-ci/rust/commit/18bebddbbdce6814465c533c7f555330b7580293/checks/27863702005/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting bootstrap v0.0.0 (/checkout/src/bootstrap)
error: unresolved link to `BootstrapCommand::capture`
  --> src/utils/exec.rs:53:61
   |
53 | /// If you want to handle the output programmatically, use [BootstrapCommand::capture].
   |                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^ the struct `BootstrapCommand` has no field or associated item named `capture`
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: could not document `bootstrap`
error: could not document `bootstrap`

Command cd "/checkout" && env -u MAKEFLAGS -u MFLAGS AR_aarch64_unknown_linux_gnu="ar" CARGO_INCREMENTAL="0" CARGO_PROFILE_RELEASE_DEBUG="0" CARGO_PROFILE_RELEASE_DEBUG_ASSERTIONS="true" CARGO_PROFILE_RELEASE_OVERFLOW_CHECKS="true" CARGO_PROFILE_RELEASE_STRIP="false" CARGO_TARGET_DIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-tools" CC_aarch64_unknown_linux_gnu="sccache cc" CFG_COMPILER_BUILD_TRIPLE="aarch64-unknown-linux-gnu" CFG_COMPILER_HOST_TRIPLE="aarch64-unknown-linux-gnu" CFG_RELEASE="1.82.0-nightly" CFG_RELEASE_CHANNEL="nightly" CFG_RELEASE_NUM="1.82.0" CFG_VERSION="1.82.0-nightly (18bebddbb 2024-07-24)" CFG_VER_DATE="2024-07-24" CFG_VER_HASH="18bebddbbdce6814465c533c7f555330b7580293" CFLAGS_aarch64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC" CXXFLAGS_aarch64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC" CXX_aarch64_unknown_linux_gnu="sccache c++" DOC_RUST_LANG_ORG_CHANNEL="https://doc.rust-lang.org/nightly" LIBC_CHECK_CFG="1" LIBRARY_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/llvm/lib" LZMA_API_STATIC="1" RANLIB_aarch64_unknown_linux_gnu="ar s" REAL_LIBRARY_PATH_VAR="LD_LIBRARY_PATH" RUSTBUILD_NATIVE_DIR="/checkout/obj/build/aarch64-unknown-linux-gnu/native" RUSTC="/checkout/obj/build/bootstrap/debug/rustc" RUSTC_BOOTSTRAP="1" RUSTC_BREAK_ON_ICE="1" RUSTC_ERROR_METADATA_DST="/checkout/obj/build/tmp/extended-error-metadata" RUSTC_HOST_FLAGS="-Zunstable-options --check-cfg=cfg(bootstrap)" RUSTC_INSTALL_BINDIR="bin" RUSTC_LIBDIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib" RUSTC_LINT_FLAGS="-Wrust_2018_idioms -Wunused_lifetimes -Dwarnings" RUSTC_PRINT_STEP_TIMINGS="1" RUSTC_REAL="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT_LIBDIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib" RUSTC_STAGE="2" RUSTC_SYSROOT="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" RUSTC_TLS_MODEL_INITIAL_EXEC="1" RUSTC_VERBOSE="0" RUSTC_WRAPPER="/checkout/obj/build/bootstrap/debug/rustc" RUSTDOC="/checkout/obj/build/bootstrap/debug/rustdoc" RUSTDOCFLAGS="--cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) -Dwarnings -Wrustdoc::invalid_codeblock_attributes --crate-version 1.82.0-nightly\t(18bebddbb\t2024-07-24) --cfg=parallel_compiler --document-private-items -Arustdoc::private-intra-doc-links --enable-index-page --show-type-layout -Zunstable-options" RUSTDOC_LIBDIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib" RUSTDOC_REAL="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" RUSTFLAGS="--cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) -Zmacro-backtrace -Csplit-debuginfo=off --cfg=parallel_compiler -Clink-args=-Wl,-z,origin -Clink-args=-Wl,-rpath,$ORIGIN/../lib -Zunstable-options -Zon-broken-pipe=kill" RUST_TEST_THREADS="8" SYSROOT="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" TERM="xterm" __CARGO_DEFAULT_LIB_METADATA="nightlytool-rustc" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "doc" "--target" "aarch64-unknown-linux-gnu" "--release" "-Zbinary-dep-depinfo" "-j" "8" "--locked" "--color" "always" "--manifest-path" "/checkout/src/bootstrap/Cargo.toml" "-Zskip-rustdoc-fingerprint" "--no-deps" "--lib" "-p" "bootstrap" (failure_mode=Exit) did not execute successfully.
Created at: src/core/builder.rs:1403:21
Executed at: src/core/build_steps/doc.rs:1043:1

Build completed unsuccessfully in 1:18:24
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10078545117/job/27868296428)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10078545117/job/27868296428","builder_name":"checks-actions"} -->

Kobzol committed:
Fix broken doc link

Kobzol commented:
Fixed broken doc link.

@bors r=onur-ozkan

bors commented:
:pushpin: Commit 17199d055964408c614adabff740cf8d44d37c2e has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 17199d055964408c614adabff740cf8d44d37c2e -->
<!-- homu: {"type":"Approved","sha":"17199d055964408c614adabff740cf8d44d37c2e","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 17199d055964408c614adabff740cf8d44d37c2e with merge 69a06000e5e17d017d616d836322f30fa2c7a261...
<!-- homu: {"type":"BuildStarted","head_sha":"17199d055964408c614adabff740cf8d44d37c2e","merge_sha":"69a06000e5e17d017d616d836322f30fa2c7a261"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc-ext`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10089414091/job/27896844818) [(plain)](https://github.com/rust-lang-ci/rust/commit/69a06000e5e17d017d616d836322f30fa2c7a261/checks/27896844818/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10089414091/job/27899270005)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10089414091/job/27899270005","builder_name":"checks-actions"} -->

Kobzol commented:
@bors retry

Spurious Windows error.

bors commented:
:hourglass: Testing commit 17199d055964408c614adabff740cf8d44d37c2e with merge 6ef5303f3bb37299a106c5e44802ccebdcab33c4...
<!-- homu: {"type":"BuildStarted","head_sha":"17199d055964408c614adabff740cf8d44d37c2e","merge_sha":"6ef5303f3bb37299a106c5e44802ccebdcab33c4"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10092630414/job/27906684563) [(plain)](https://github.com/rust-lang-ci/rust/commit/6ef5303f3bb37299a106c5e44802ccebdcab33c4/checks/27906684563/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
thread 'main' panicked at src\utils\exec.rs:277:33:
Accessing stdout of a command that did not capture stdout
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 1:22:22
make: *** [Makefile:102: ci-msvc-ps1] Error 1
  network time: Thu, 25 Jul 2024 12:07:29 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10092630414/job/27910253454)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10092630414/job/27910253454","builder_name":"checks-actions"} -->

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit 59803286baf35cfab1d3cd6085f634d775e454bf with merge 597741d1c012edca0b575d27db5f969c7f08113b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"59803286baf35cfab1d3cd6085f634d775e454bf","merge_sha":"597741d1c012edca0b575d27db5f969c7f08113b"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10101038067/job/27933724867) [(plain)](https://github.com/rust-lang-ci/rust/commit/597741d1c012edca0b575d27db5f969c7f08113b/checks/27933724867/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  99% (48331/48802)
Updating files:  99% (48747/48802)
Updating files: 100% (48802/48802)
Updating files: 100% (48802/48802), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'597741d1c012edca0b575d27db5f969c7f08113b'
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
   2: core::panicking::panic_display
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library\core\src\panicking.rs:262
   3: core::option::expect_failed
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library\core\src\option.rs:1995
   4: core::option::Option::expect<alloc::vec::Vec<u8,alloc::alloc::Global> >
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\core\src\option.rs:898
   5: bootstrap::utils::exec::CommandOutput::stdout
             at .\src\bootstrap\src\utils\exec.rs:277
   6: bootstrap::core::build_steps::test::get_browser_ui_test_version_inner
             at .\src\bootstrap\src\core\build_steps\test.rs:912
   7: bootstrap::core::build_steps::test::get_browser_ui_test_version
             at .\src\bootstrap\src\core\build_steps\test.rs:920
   8: bootstrap::core::build_steps::test::impl$15::should_run::closure$0::closure$0
             at .\src\bootstrap\src\core\build_steps\test.rs:945
   9: core::option::Option::map<ref$<std::path::PathBuf>,bool,bootstrap::core::build_steps::test::impl$15::should_run::closure$0::closure_env$0>
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\core\src\option.rs:1075
  10: bootstrap::core::build_steps::test::impl$15::should_run::closure$0
             at .\src\bootstrap\src\core\build_steps\test.rs:941
  11: core::ops::function::FnOnce::call_once<bootstrap::core::build_steps::test::impl$15::should_run::closure_env$0,tuple$<> >
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\core\src\ops\function.rs:250
  12: core::ops::function::FnOnce::call_once<bootstrap::core::build_steps::test::impl$15::should_run::closure_env$0,tuple$<> >
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\core\src\ops\function.rs:250
  13: alloc::boxed::impl$48::call_once<tuple$<>,dyn$<core::ops::function::Fn<tuple$<>,assoc$<Output,bool> > >,alloc::alloc::Global>
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\alloc\src\boxed.rs:2063
  14: std::sync::lazy_lock::impl$0::force::closure$0<bool,alloc::boxed::Box<dyn$<core::ops::function::Fn<tuple$<>,assoc$<Output,bool> > >,alloc::alloc::Global> >
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\std\src\sync\lazy_lock.rs:161
  15: std::sync::once::impl$2::call_once::closure$0<std::sync::lazy_lock::impl$0::force::closure_env$0<bool,alloc::boxed::Box<dyn$<core::ops::function::Fn<tuple$<>,assoc$<Output,bool> > >,alloc::alloc::Global> > >
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\std\src\sync\once.rs:149
  16: std::sys::sync::once::queue::Once::call
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library\std\src\sys\sync\once\queue.rs:183
  17: std::sync::once::Once::call_once<std::sync::lazy_lock::impl$0::force::closure_env$0<bool,alloc::boxed::Box<dyn$<core::ops::function::Fn<tuple$<>,assoc$<Output,bool> > >,alloc::alloc::Global> > >
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\std\src\sync\once.rs:149
  18: std::sync::lazy_lock::LazyLock::force
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\std\src\sync\lazy_lock.rs:157
  19: std::sync::lazy_lock::impl$3::deref<bool,alloc::boxed::Box<dyn$<core::ops::function::Fn<tuple$<>,assoc$<Output,bool> > >,alloc::alloc::Global> >
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\std\src\sync\lazy_lock.rs:215
  20: bootstrap::core::builder::ShouldRun::is_really_default
             at .\src\bootstrap\src\core\builder.rs:570
  21: bootstrap::core::builder::StepDescription::run
             at .\src\bootstrap\src\core\builder.rs:432
  22: bootstrap::core::builder::Builder::run_step_descriptions
             at .\src\bootstrap\src\core\builder.rs:1101
             at .\src\bootstrap\src\core\builder.rs:1081
  24: bootstrap::Build::build
             at .\src\bootstrap\src\lib.rs:662
  25: bootstrap::main
  25: bootstrap::main
             at .\src\bootstrap\src\bin\main.rs:85
  26: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\core\src\ops\function.rs:250
             at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a\library\core\src\hint.rs:338
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Build completed unsuccessfully in 1:23:06
Build completed unsuccessfully in 1:23:06
make: *** [Makefile:102: ci-msvc-ps1] Error 1
  network time: Thu, 25 Jul 2024 22:05:59 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10101038067/job/27937278662)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10101038067/job/27937278662","builder_name":"checks-actions"} -->

Kobzol committed:
Fix storing of stdout/stderr in bootstrap commands that failed to start

Before, their stdout/stderr was forcefully set to `None`, even if the corresponding command tried to capture output.

onur-ozkan commented on src/bootstrap/src/utils/exec.rs:
```diff
@@ -230,17 +223,41 @@ pub fn command<S: AsRef<OsStr>>(program: S) -> BootstrapCommand {
     #[must_use]
-    pub fn did_not_start() -> Self {
-        Self { status: CommandStatus::DidNotStart, stdout: vec![], stderr: vec![] }
+    pub fn did_not_start(stdout: OutputMode, stderr: OutputMode) -> Self {
+        Self {
+            status: CommandStatus::DidNotStart,
+            stdout: match stdout {
+                OutputMode::Print => None,
+                OutputMode::Capture => Some(vec![]),
+            },
+            stderr: match stderr {
+                OutputMode::Print => None,
+                OutputMode::Capture => Some(vec![]),
+            },
+        }
+    }
```
If command isn't started, why do we need to pass these arguments? Can't we handle this with some default values?

Kobzol replied:
The output mode determines if it is valid to ask for stdout/stderr. E.g. you can execute a command without capturing its stderr, and then try to read the stderr. In that case it's a bug (the code shouldn't just silently return an empty string), and we need to propagate the information whether the output was captured or not so that we're able to find the bug (rather than just returning an empty string).

onur-ozkan replied:
> The output mode determines if it is valid to ask for stdout/stderr.

Function is called `did_not_start`, which makes me think that we shouldn't need to ask for stdout/stderr. Am I missing something here?

Kobzol replied:
The command can be used with `.allow_failure()`. So it doesn't matter if it failed to start or failed to execute, if it fails, and the code uses `allow_failure()`, it is able to inspect the result of the command (including stdout/stderr).

We could model this in a different way (e.g. returning `Result` to signify failures of command start/execution), but that would require a large change across bootstrap to handle the potential failures.

onur-ozkan replied:
What would go wrong if we return `Some(vec![])` for both stdout and stderr from `did_not_start` without requiring any argument?

Kobzol replied:
Nothing much really, just that if there were some code that does `run_capture_stdout().stderr()`, and it failed to start, the code wouldn't panic (to tell us the bug).

onur-ozkan commented:
@bors r+

bors commented:
:pushpin: Commit abd8768768cf47829b8f9826e69d083657e8b8fd has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan abd8768768cf47829b8f9826e69d083657e8b8fd -->
<!-- homu: {"type":"Approved","sha":"abd8768768cf47829b8f9826e69d083657e8b8fd","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit abd8768768cf47829b8f9826e69d083657e8b8fd with merge 1b51d80027919563004918eaadfa0d890ac0eb93...
<!-- homu: {"type":"BuildStarted","head_sha":"abd8768768cf47829b8f9826e69d083657e8b8fd","merge_sha":"1b51d80027919563004918eaadfa0d890ac0eb93"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10131392828/job/28016095453)
Approved by: onur-ozkan
Pushing 1b51d80027919563004918eaadfa0d890ac0eb93 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"onur-ozkan","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10131392828/job/28016095453"},"merge_sha":"1b51d80027919563004918eaadfa0d890ac0eb93"} -->

rust-timer commented:
Finished benchmarking commit (1b51d80027919563004918eaadfa0d890ac0eb93): [comparison URL](https://perf.rust-lang.org/compare.html?start=3954398882707ce3f683722795d1a7111312b5d9&end=1b51d80027919563004918eaadfa0d890ac0eb93&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=3954398882707ce3f683722795d1a7111312b5d9&end=1b51d80027919563004918eaadfa0d890ac0eb93&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3954398882707ce3f683722795d1a7111312b5d9&end=1b51d80027919563004918eaadfa0d890ac0eb93&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=3954398882707ce3f683722795d1a7111312b5d9&end=1b51d80027919563004918eaadfa0d890ac0eb93&stat=cycles:u)
<details>
<summary>Results (secondary -2.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-2.7%, -2.7%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3954398882707ce3f683722795d1a7111312b5d9&end=1b51d80027919563004918eaadfa0d890ac0eb93&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 770.349s -> 770.165s (-0.02%)
**Artifact size**: 331.55 MiB -> 331.66 MiB (0.03%)
